falcon/hooks.py

```diff
@@ -14,21 +14,35 @@
 
 """Hook decorators."""
 
+from __future__ import annotations
+
 from functools import wraps
 from inspect import getmembers
 from inspect import iscoroutinefunction
 import re
+import typing as t
 
 from falcon.constants import COMBINED_METHODS
 from falcon.util.misc import get_argnames
 from falcon.util.sync import _wrap_non_coroutine_unsafe
 
+if t.TYPE_CHECKING:  # pragma: no cover
+    import falcon as wsgi
+    from falcon import asgi
+
 _DECORABLE_METHOD_NAME = re.compile(
     r'^on_({})(_\w+)?$'.format('|'.join(method.lower() for method in COMBINED_METHODS))
 )
 
+Resource = object
```

Comment:
I think it would make sense to consolidate these kind of things in `typing.py`, that's why I created that module :slightly_smiling_face: 
But let's do that in follow-up PRs.